scipy/spatial/ckdtree.pyx

```diff
@@ -340,52 +344,70 @@ cdef class cKDTreeNode:
+        self._subtree_indices = None
+        if self.split_dim == -1:
+            self.lesser = None
+            self.greater = None
+        else:
+            # setup lesser branch
+            n1 = cKDTreeNode()
+            n1._node = self._node.less
+            n1._data = self._data
+            n1._indices = self._indices
+            n1.level = self.level + 1
+            n1._setup()
+            self.lesser = n1
+            # setup greater branch
+            n2 = cKDTreeNode()
+            n2._node = self._node.greater
+            n2._data = self._data
+            n2._indices = self._indices
+            n2.level = self.level + 1
+            n2._setup()
+            self.greater = n2
 
     property data_points:
         def __get__(cKDTreeNode self):
             return self._data[self.indices,:]
 
     property indices:
         def __get__(cKDTreeNode self):
-            cdef np.intp_t i, start, stop
-            if self.split_dim == -1:
-                start = self._node.start_idx
-                stop = self._node.end_idx
-                return self._indices[start:stop]
-            else:
-                return np.hstack([self.lesser.indices,
-                           self.greater.indices])
-
-    property lesser:
-        def __get__(cKDTreeNode self):
-            if self.split_dim == -1:
-                return None
-            else:
-                n = cKDTreeNode()
-                n._node = self._node.less
-                n._data = self._data
-                n._indices = self._indices
-                n.level = self.level + 1
-                n._setup()
-                return n
-
-    property greater:
-        def __get__(cKDTreeNode self):
-            if self.split_dim == -1:
-                return None
-            else:
-                n = cKDTreeNode()
-                n._node = self._node.greater
-                n._data = self._data
-                n._indices = self._indices
-                n.level = self.level + 1
-                n._setup()
-                return n
+            cdef: 
+                np.intp_t s0, s1, s2
+                np.intp_t *d0
+                np.intp_t *d1
+                np.intp_t *d2
+                np.ndarray i0
+                np.ndarray i2
+                np.ndarray i3
+            if self._subtree_indices is None:
+                if self.split_dim == -1:
+                    start = self._node.start_idx
+                    stop = self._node.end_idx
+                    self._subtree_indices = self._indices[start:stop]
+                else:
+                    i1 = self.lesser.indices
+                    i2 = self.greater.indices
+                    s1 = np.PyArray_DIMS(i1)[0]
+                    d1 = <np.intp_t*> np.PyArray_DATA(i1)
+                    s2 = np.PyArray_DIMS(i2)[0]
+                    d2 = <np.intp_t*> np.PyArray_DATA(i2)
+                    s0 = s1 + s2
+                    i0 = np.PyArray_SimpleNew(1, &s0, np.NPY_INTP)
```

Comment:
I don't think the problem was `hstack`. This algorithm is just terribly inefficient because it copies the data around so many times and creates O(n logn) new arrays. Instead, it should build a single array starting from the root node. You traverse the tree in left to right order and store the indices such than each node is over a contiguous range, even the inner nodes. Then `node._subtree_indices` is just a view into one shared indices array.

Reply:
Yes, start from the indices array at the root and shop it up as we move up to the leaves, instead of starting at the leaves and gathering multiple indices arrays as we move downwards. That would certainly be a lot better. Better algorithm and it requires far less storage space.

Reply:
Actually this is silly. ```self._indices``` is already sorted, so it just boils down to this:
```
property indices:
        def __get__(cKDTreeNode self):
            cdef:
                np.intp_t start
                np.intp_t stop
            start = self._node.start_idx
            stop = self._node.end_idx
            return self._indices[start:stop] 
```